includes/class-amp-service-worker.php

```diff
@@ -0,0 +1,316 @@
+					),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Add runtime caching of Google Fonts with stale-while-revalidate strategy for stylesheets and cache-first strategy for webfont files.
+	 *
+	 * @link https://developers.google.com/web/tools/workbox/guides/common-recipes#google_fonts
+	 * @link https://github.com/ampproject/amp-by-example/blob/4593af61609898043302a101826ddafe7206bfd9/boilerplate-generator/templates/files/serviceworkerJs.js#L76-L103
+	 * @link https://github.com/xwp/pwa-wp/blob/master/integrations/class-wp-service-worker-fonts-integration.php
+	 *
+	 * @param WP_Service_Worker_Scripts $service_workers Service workers.
+	 */
+	public static function add_google_fonts_caching( $service_workers ) {
+
+		// The PWA plugin also automatically adds runtime caching for Google Fonts when WP_SERVICE_WORKER_INTEGRATIONS_ENABLED is set.
+		if ( class_exists( 'WP_Service_Worker_Fonts_Integration' ) ) {
+			return;
+		}
+
+		// Cache the Google Fonts stylesheets with a stale while revalidate strategy.
+		$service_workers->caching_routes()->register(
+			'^https:\/\/fonts\.googleapis\.com',
+			array(
+				'strategy'  => WP_Service_Worker_Caching_Routes::STRATEGY_STALE_WHILE_REVALIDATE,
+				'cacheName' => 'google-fonts-stylesheets',
+			)
+		);
+
+		// Cache the Google Fonts webfont files with a cache first strategy for 1 year.
+		$service_workers->caching_routes()->register(
+			'^https:\/\/fonts\.gstatic\.com',
+			array(
+				'strategy'  => WP_Service_Worker_Caching_Routes::STRATEGY_CACHE_FIRST,
+				'cacheName' => 'google-fonts-webfonts',
+				'plugins'   => array(
+					'cacheableResponse' => array(
+						'statuses' => array( 0, 200 ),
+					),
+					'expiration'        => array(
+						'maxAgeSeconds' => YEAR_IN_SECONDS,
+						'maxEntries'    => 30,
+					),
+				),
+			)
+		);
+	}
+
+	/**
+	 * Register URLs that will be precached in the runtime cache. (Yes, this sounds somewhat strange.)
+	 *
+	 * Note that the PWA plugin handles the precaching of custom logo, custom header,
+	 * and custom background. The PWA plugin also handles precaching & serving of the
+	 * offline/500 error pages and enabling navigation preload.
+	 *
+	 * @link https://github.com/ampproject/amp-by-example/blob/4593af61609898043302a101826ddafe7206bfd9/boilerplate-generator/templates/files/serviceworkerJs.js#L9-L22
+	 * @see AMP_Service_Worker::add_cdn_script_caching()
+	 *
+	 * @return array Runtime pre-cached URLs.
+	 */
+	public static function get_precached_script_cdn_urls() {
+
+		// List of AMP scripts that we know will be used in WordPress always.
+		$precached_handles = array(
+			'amp-runtime',
+			'amp-bind', // Used by comments.
+			'amp-form', // Used by comments.
+			'amp-install-serviceworker',
+		);
+
+		$theme_support = AMP_Theme_Support::get_theme_support_args();
+		if ( ! empty( $theme_support['comments_live_list'] ) ) {
+			$precached_handles[] = 'amp-live-list';
+		}
+
+		if ( amp_get_analytics() ) {
+			$precached_handles[] = 'amp-analytics';
+		}
```

Comment:
Couldn't we cover other AMP extensions here as well? What would be the downsides of including them in this list unconditionally?

Alternatively, we could at least make the list filterable.

Reply:
The downside would be needlessly precaching AMP scripts and wasting the user's storage.

The intention here is that the service worker can automatically precache scripts that we know will be used purely based on configuration. If a theme wants to precache other scripts then they could do so via:

```php
add_action(
	'wp_front_service_worker',
	function ( WP_Service_Worker_Scripts $sw_scripts ) {
		$sw_scripts->precaching_routes()->register( wp_scripts()->registered['amp-carousel']->src );
	}
);
```

Alternatively, we could include a `precached_scripts` in the theme support args which takes a list of script handles, which would be more user friendly.

Reply:
Makes sense.

I think applying a filter here would be a good start. I'm not sure we should add more complexity to the theme support arguments. For example, these precached scripts should be provided somewhere under the service_worker key of AMP theme support, but this would cause for an even deeper nested array.

Having a filter allows advanced users to adjust this, and then in the future we can figure out how to make this more straightforward, or whether we can possibly auto-detect it somehow.

Reply:
How about deferring this decision for later? Themes can use the `wp_front_service_worker` hook to accomplish this for now. We can then determine later what level of abstraction is right based on usage.